File: melosys-eessi-app/src/main/java/no/nav/melosys/eessi/service/sed/mapper/fra_sed/melosys_eessi_melding/MelosysEessiMeldingMapperA009.java
package no.nav.melosys.eessi.service.sed.mapper.fra_sed.melosys_eessi_melding;

import java.time.LocalDate;

import lombok.extern.slf4j.Slf4j;
import no.nav.melosys.eessi.models.sed.SED;
import no.nav.melosys.eessi.models.sed.medlemskap.impl.MedlemskapA009;
import no.nav.melosys.eessi.models.sed.nav.AapenPeriode;
import no.nav.melosys.eessi.models.sed.nav.Fastperiode;
import no.nav.melosys.eessi.models.sed.nav.Periode;

import static no.nav.melosys.eessi.models.DatoUtils.tilLocalDate;

@Slf4j
class MelosysEessiMeldingMapperA009 implements NyttLovvalgEessiMeldingMapper<MedlemskapA009> {

    @Override
    public Boolean sedErEndring(MedlemskapA009 medlemskap) {
        var erEndring = !"ja".equalsIgnoreCase(medlemskap.getVedtak().getEropprinneligvedtak());
        log.info("sedErEndring i A009 er {}, med erendringsvedtak: {} og eropprinneligvedtak: {}", erEndring,
            medlemskap.getVedtak().getErendringsvedtak(), medlemskap.getVedtak().getEropprinneligvedtak());
        return erEndring;
    }

    @Override
    public MedlemskapA009 hentMedlemskap(SED sed) {
        return (MedlemskapA009) sed.getMedlemskap();
    }

    @Override
    public String hentLovvalgsbestemmelse(MedlemskapA009 medlemskap) {
        return medlemskap.getVedtak().getArtikkelforordning();
    }

    @Override
    public String hentLovvalgsland(MedlemskapA009 medlemskap) {
        return medlemskap.getVedtak().getLand();
    }

    @Override
    public no.nav.melosys.eessi.kafka.producers.model.Periode mapPeriode(MedlemskapA009 medlemskap) {

        LocalDate fom;
        LocalDate tom;

        Periode periode = medlemskap.getVedtak().getGjelderperiode();
        if (periode.erAapenPeriode()) {
            AapenPeriode aapenPeriode = periode.getAapenperiode();
            fom = tilLocalDate(aapenPeriode.getStartdato());
            tom = null;
        } else {
            Fastperiode fastperiode = periode.getFastperiode();
            fom = tilLocalDate(fastperiode.getStartdato());
            tom = tilLocalDate(fastperiode.getSluttdato());
        }

        return new no.nav.melosys.eessi.kafka.producers.model.Periode(fom, tom);
    }
}

File: melosys-eessi-app/src/main/java/no/nav/melosys/eessi/service/sed/mapper/fra_sed/melosys_eessi_melding/MelosysEessiMeldingMapperA009.kt
package no.nav.melosys.eessi.service.sed.mapper.fra_sed.melosys_eessi_melding

import mu.KotlinLogging
import no.nav.melosys.eessi.kafka.producers.model.Periode
import no.nav.melosys.eessi.models.DatoUtils.tilLocalDate
import no.nav.melosys.eessi.models.sed.SED
import no.nav.melosys.eessi.models.sed.medlemskap.impl.MedlemskapA009
import java.time.LocalDate

private val log = KotlinLogging.logger { }

internal class MelosysEessiMeldingMapperA009 : NyttLovvalgEessiMeldingMapper<MedlemskapA009> {
    override fun sedErEndring(medlemskap: MedlemskapA009): Boolean {
        val erEndring = !"ja".equals(medlemskap.vedtak!!.eropprinneligvedtak, ignoreCase = true)
        log.info(
            "sedErEndring i A009 er $erEndring, med erendringsvedtak: ${
                medlemskap.vedtak!!.erendringsvedtak
            } og eropprinneligvedtak: ${medlemskap.vedtak!!.eropprinneligvedtak}",
        )
        return erEndring
    }

    override fun hentMedlemskap(sed: SED): MedlemskapA009 = sed.medlemskap as MedlemskapA009

    override fun hentLovvalgsbestemmelse(medlemskap: MedlemskapA009): String? = medlemskap.vedtak!!.artikkelforordning

    override fun hentLovvalgsland(medlemskap: MedlemskapA009): String? = medlemskap.vedtak!!.land

    override fun mapPeriode(medlemskap: MedlemskapA009): Periode {
        val fom: LocalDate
        val tom: LocalDate?
        val periode = medlemskap.vedtak!!.gjelderperiode
        if (periode!!.erAapenPeriode()) {
            val aapenPeriode = periode.aapenperiode
            fom = tilLocalDate(aapenPeriode!!.startdato!!)
            tom = null
        } else {
            val fastperiode = periode.fastperiode
            fom = tilLocalDate(fastperiode!!.startdato!!)
            tom = tilLocalDate(fastperiode.sluttdato!!)
        }
        return Periode(fom, tom)
    }
}
